### PR TITLE
ENH: Arbiter Request Sorting and Filtering

### PR DESCRIPTION
**Pull request by ZLLentz**
Continuation of #35 to add filtering and sorting to the pre-emptive requests table. In the process I did a bunch of refactoring and large tweaks, though the core setup that Cristina used is intact.

This allows you to sort on any of the table fields and to filter out the noisy/useless entries. These filters are on by default and I suspect they will only be turned off for system debug.

The table will sort upon any new selection, and then again when you click "sort again". If you click "auto update" the table will sort again when new values are read from PVs.

Checklist:
- [x] Figure out what is here now from Cristina and understand how it works
- [x] Make sure we support sorting on at least one field at a time
- [x] Make sure it is unambiguous what exactly we are sorting by
- [x] Sort by allowed beam rate
- [x] Sort by allowed transmission
- [x] Sort by number of eV ranges allowed
- [x] Sort by assertion ID
- [x] Sort by the device name field
- [x] Unsorted (original sorting by config file)
- [x] Optional auto-update of sort as values change
- [x] Option to hide lines that have no beam restrictions
- [x] Option to hide lines that are inactive
- [x] Option to hide lines that are disconnected
- [x] bugfixes and testing
- [x] layout nitpicking
- [ ] bugfixes and testing while moving real devices

Preview with default setup upon opening the screen. Note how it's now much easier to understand what is going on:

image

I think I might backport this to the branch that LFE is running on because it shouldn't be too hard and will be useful.

### CODE REVIEW

**Review by ghalym — APPROVED** [review]

File: preemptive_requests.py [hunk 1]
```diff
@@ -35,23 +77,362 @@ def setup_requests(self):
...
+        return self.get_value() < other.get_value()
+
+
+def bitmask_count(bitmask):
+    """Count the number of high bits in a bitmask."""
+    bitmask = int(bitmask)
+    count = 0
+    if bitmask < 0:
+        bitmask += 2**32
+    while bitmask > 0:
```
Comment (klauer) [inline preemptive_requests.py:343]: Lazy me sees this and thinks, hmm... `str(bin(bitmask)).count('1')`

File: preemptive_requests.py [hunk 2]
```diff
@@ -35,23 +77,362 @@ def setup_requests(self):
...
+    """Convert an EPICS char waveform to a str."""
+    text = ''
+    for num in waveform_array:
+        if num == 0:
+            break
+        text += chr(num)
+    return text
+
+
+@dataclass(frozen=True)
```
Comment (klauer) [inline preemptive_requests.py:360]: YES! Dataclasses!

File: preemptive_requests.py [hunk 3]
```diff
@@ -35,23 +77,362 @@ def setup_requests(self):
...
+        Function or type to call on the str stored in the table before
+        comparing with other PMPSTableWidgetItem instances for sorting.
+    default : Any
+        A starting value for the widget item.
+    channel : str, optional
+        PyDM channel address for value and connection updates.
+    """
+    def __init__(self, store_type, data_type, default,
+                 channel=None, parent=None):
+        QTableWidgetItem.__init__(self, parent)
```
Comment (klauer) [inline preemptive_requests.py:295]:
```suggestion
        super().__init__(parent)
```

File: preemptive_requests.py [hunk 4]
```diff
@@ -1,27 +1,69 @@
...
+    - Visibility of an item in a QTableWidget is not controlled by the normal
+      visibility properties, so the visibility rules cannot be used the same
+      way as in the fast faults tab. When you try this, you just see every row
+      of the table all the time. Therefore, the previous approch is not useful
+      here, and instead we link the channels up with the QTableWidget.hideRow
+      and QTableWidget.showRow methods.
+    """
     def __init__(self, parent=None, args=None, macros=None):
-        super(PreemptiveRequests, self).__init__(parent=parent, args=args, macros=macros)
+        super(PreemptiveRequests, self).__init__(parent=parent,
```
Comment (klauer) [inline preemptive_requests.py:40]:
```suggestion
        super().__init__(parent=parent,
```

**Review by klauer — APPROVED** [review]
Proof is in the pudding, so they say - and the GUI works well.

Just some minor code suggestions as I read through the PR.

File: preemptive_requests.py (hunk 1 above)
Comment (ZLLentz) [reply preemptive_requests.py:343]: Ah! This is what happens when I code in non-python, I forget that there's a more straighforward way to do everything.

File: preemptive_requests.py (hunk 2 above)
Comment (ZLLentz) [reply preemptive_requests.py:360]: Yes! Gotta use all the new toys

File: preemptive_requests.py (hunk 3 above)
Comment (ZLLentz) [reply preemptive_requests.py:295]: Good catch, this and the other one must be leftover from Hugo and/or Cristina where they were forced to keep pydm code python2 compatible...

**Comment by ZLLentz** [comment]
I'm going to deploy this either tonight or tomorrow and do further updates in a follow-up.

**Pull request merged**